### PR TITLE
WIP:  Resolve "Panzer:  Update Integrator_DivBasisTimesScalar"

### LINKED ISSUE #1891: Panzer:  Update Integrator_DivBasisTimesScalar
The `panzer::Integrator_DivBasisTimesScalar` class needs to be updated to support the new `CONTRIBUTES` style `Evaluator`, as per #1575.  An example of the changes to be made can be found in the `panzer::Integrator_BasisTimesScalar` class.
@trilinos/panzer

### PR DESCRIPTION
**Pull request by jmgate**
@trilinos/panzer 

## Description
Refactors `panzer::Integrator_DivBasisTimesScalar` such that it now has a compile-time-argument-checking constructor which enables its use as a `CONTRIBUTES`-style `Evaluator`.  Also Kokkos-ifies the `evaluateFields()` routine.

## Related Issues
Closes #1891 

## Motivation and Context
Memory savings&mdash;see #1575.

## How Has This Been Tested?
Panzer/Drekar/Charon tests pass on my RHEL7 machine.

## Checklist

- [ ] My commit messages mention the appropriate GitHub issue numbers.
- [x] My code follows the code style of the affected package(s).
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] No new compiler warnings were introduced.
- [ ] These changes break backwards compatibility.

### CODE REVIEW

**Comment by jmgate** [comment]
@rppawlo, the changes to the `Integrator_DivBasisTimesScalar` class are extensive.  The substantive commits to review are ed4edcb, in which a main, compile-time-argument-checking constructor was created, along with a delegating `ParameterList` constructor, and then cda0fd8, in which `evaluateFields()` was rewritten to use `operator()()`.  All other commits involve formatting, naming conventions, Doxygen, etc.

If this looks good to you, I'll squash everything before the merge.

**Comment by jmgate** [comment]
> I think we just need to make sure to test on cuda with the non-zero number of field multipliers.

I'm afraid the only tests of this class we have in Panzer's suite don't include any field multipliers.  The implementation is identical is identical to that for `Integrator_BasisTimesScalar`, which is tested with multiple field multipliers.  Is that sufficient demonstration, or should we add another test before merging this?

File: packages/panzer/disc-fe/src/evaluators/Panzer_Integrator_DivBasisTimesScalar_impl.hpp [hunk 1]
```diff
@@ -40,151 +40,289 @@
...
+    PHX::FieldManager<Traits>& fm)
+  {
+    using Kokkos::createDynRankView;
+    using panzer::getBasisIndex;
 
-  this->setName(n);
-}
+    // Get the Kokkos::Views of the field multipliers.
+    for (size_t i(0); i < fieldMults_.size(); ++i)
+      kokkosFieldMults_(i) = fieldMults_[i].get_static_view();
```
[the remark this answers is not in the document]
Comment (jmgate) [reply packages/panzer/disc-fe/src/evaluators/Panzer_Integrator_DivBasisTimesScalar_impl.hpp]: Done.  What exactly does the `fence` do there for us?

**Comment by jmgate** [comment]
@rppawlo, I added the `fence` and squashed.  I think you'll need to re-approve.

File: packages/panzer/disc-fe/src/evaluators/Panzer_Integrator_DivBasisTimesScalar_impl.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (rppawlo) [reply packages/panzer/disc-fe/src/evaluators/Panzer_Integrator_DivBasisTimesScalar_impl.hpp]: That protects us from accidentally using the View of views before the uvm has completed copying things over. In the kokkos manual, they recommended this for our use case of building the outer view with UVM. Just a harmless safety check.
